File: src/server/routes/text2sql.py
# Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
# SPDX-License-Identifier: MIT

"""
Text2SQL API routes for DeerFlow.

Provides REST API endpoints for Text2SQL functionality including SQL generation,
training data management, and real-time WebSocket connections.
"""

import logging
from typing import List, Optional, Dict, Any
from fastapi import APIRouter, HTTPException, WebSocket, Query, Path, Body

from src.services.text2sql import Text2SQLService
from src.models.text2sql import (
    QueryStatus,
    TrainingDataType,
    SQLGenerationRequest,
    SQLGenerationResponse,
    SQLExecutionRequest,
    SQLExecutionResponse,
    TrainingDataRequest,
    TrainingDataResponse,
    Text2SQLStatistics,
    QuestionAnswerRequest,
    QuestionAnswerResponse,
    BatchTrainingRequest
)
from src.websocket.text2sql_manager import handle_websocket_connection

logger = logging.getLogger(__name__)

# Create router
router = APIRouter(prefix="/api/text2sql", tags=["Text2SQL"])

# Initialize service
text2sql_service = Text2SQLService()


# SQL Generation and Execution Endpoints

@router.post("/generate", response_model=SQLGenerationResponse)
async def generate_sql(request: SQLGenerationRequest):
    """Generate SQL from natural language question"""
    try:
        response = await text2sql_service.generate_sql(request)
        return response
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to generate SQL: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/execute", response_model=SQLExecutionResponse)
async def execute_sql(request: SQLExecutionRequest):
    """Execute generated SQL query"""
    try:
        response = await text2sql_service.execute_sql(request)
        return response
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to execute SQL: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


# Query History Endpoints

@router.get("/history")
async def get_query_history(
    datasource_id: Optional[int] = Query(None, description="Filter by datasource ID"),
    status: Optional[QueryStatus] = Query(None, description="Filter by query status"),
    limit: int = Query(50, ge=1, le=100, description="Number of records to return"),
    offset: int = Query(0, ge=0, description="Number of records to skip")
):
    """Get query history with pagination"""
    try:
        queries, total = await text2sql_service.get_query_history(
            datasource_id=datasource_id,
            status=status,
            limit=limit,
            offset=offset
        )
        
        return {
            "queries": [query.model_dump() for query in queries],
            "total": total,
            "limit": limit,
            "offset": offset
        }
    except Exception as e:
        logger.error(f"Failed to get query history: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


# Training Data Endpoints

@router.post("/training", response_model=TrainingDataResponse)
async def add_training_data(request: TrainingDataRequest):
    """Add training data for model improvement"""
    try:
        response = await text2sql_service.add_training_data(request)
        return response
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to add training data: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.get("/training")
async def get_training_data(
    datasource_id: Optional[int] = Query(None, description="Filter by datasource ID"),
    content_type: Optional[TrainingDataType] = Query(None, description="Filter by content type"),
    is_active: bool = Query(True, description="Filter by active status"),
    limit: int = Query(50, ge=1, le=100, description="Number of records to return"),
    offset: int = Query(0, ge=0, description="Number of records to skip")
):
    """Get training data with pagination"""
    try:
        training_data, total = await text2sql_service.get_training_data_list(
            datasource_id=datasource_id,
            content_type=content_type,
            is_active=is_active,
            limit=limit,
            offset=offset
        )
        
        return {
            "training_data": [data.model_dump() for data in training_data],
            "total": total,
            "limit": limit,
            "offset": offset
        }
    except Exception as e:
        logger.error(f"Failed to get training data: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.put("/training/{training_id}", response_model=TrainingDataResponse)
async def update_training_data(
    training_id: int = Path(..., description="Training data ID"),
    request: TrainingDataRequest = Body(..., description="Updated training data")
):
    """Update training data"""
    try:
        # Get existing training data
        existing_data = await text2sql_service.repository.get_training_data(training_id)
        if not existing_data:
            raise HTTPException(status_code=404, detail="Training data not found")

        # Update the training data
        updated = await text2sql_service.repository.update_training_data(
            training_id=training_id,
            content_type=request.content_type,
            content=request.content,
            question=request.question,
            sql_query=request.sql_query,
            table_names=request.table_names,
            database_schema=request.database_schema,
            metadata=request.metadata
        )

        if not updated:
            raise HTTPException(status_code=404, detail="Training data not found")

        # Get updated training data
        updated_data = await text2sql_service.repository.get_training_data(training_id)

        # Trigger embedding regeneration for updated data
        try:
            await text2sql_service.generate_embeddings_for_training_data(
                datasource_id=updated_data.datasource_id,
                training_data_ids=[training_id]
            )
            logger.info(f"Started embedding regeneration for updated training data {training_id}")
        except Exception as e:
            logger.warning(f"Failed to start embedding regeneration for training data {training_id}: {e}")

        return TrainingDataResponse(
            id=updated_data.id,
            datasource_id=updated_data.datasource_id,
            content_type=updated_data.content_type,
            question=updated_data.question,
            sql_query=updated_data.sql_query,
            content=updated_data.content,
            table_names=updated_data.table_names,
            database_schema=updated_data.database_schema,
            metadata=updated_data.metadata,
            content_hash=updated_data.content_hash,
            is_active=updated_data.is_active,
            is_validated=updated_data.is_validated,
            validation_score=updated_data.validation_score,
            created_at=updated_data.created_at,
            updated_at=updated_data.updated_at
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to update training data: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.delete("/training/{training_id}")
async def delete_training_data(
    training_id: int = Path(..., description="Training data ID"),
    soft_delete: bool = Query(True, description="Perform soft delete")
):
    """Delete training data"""
    try:
        deleted = await text2sql_service.delete_training_data(training_id, soft_delete)
        if not deleted:
            raise HTTPException(status_code=404, detail="Training data not found")

        return {"message": "Training data deleted successfully", "training_id": training_id}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to delete training data: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


# Training Session and Model Management Endpoints

@router.post("/retrain/{datasource_id}")
async def retrain_model(
    datasource_id: int = Path(..., description="Datasource ID"),
    force_rebuild: bool = Query(False, description="Force rebuild all embeddings")
):
    """Start model retraining for a datasource"""
    try:
        task_id = await text2sql_service.retrain_model(datasource_id, force_rebuild)
        return {
            "message": "Model retraining started",
            "task_id": task_id,
            "datasource_id": datasource_id,
            "force_rebuild": force_rebuild
        }
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to start retraining: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/training-session")
async def start_training_session(
    datasource_id: int = Body(..., description="Datasource ID"),
    session_name: Optional[str] = Body(None, description="Session name"),
    model_version: Optional[str] = Body(None, description="Model version"),
    training_parameters: Optional[Dict[str, Any]] = Body(None, description="Training parameters"),
    notes: Optional[str] = Body(None, description="Session notes")
):
    """Start a new training session"""
    try:
        result = await text2sql_service.start_training_session(
            datasource_id=datasource_id,
            session_name=session_name,
            model_version=model_version,
            training_parameters=training_parameters,
            notes=notes
        )
        return result
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to start training session: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.get("/training-sessions")
async def get_training_sessions(
    datasource_id: Optional[int] = Query(None, description="Filter by datasource ID"),
    limit: int = Query(50, ge=1, le=100, description="Number of records to return"),
    offset: int = Query(0, ge=0, description="Number of records to skip")
):
    """Get training sessions with pagination"""
    try:
        sessions, total = await text2sql_service.list_training_sessions(
            datasource_id=datasource_id,
            limit=limit,
            offset=offset
        )
        
        return {
            "sessions": [session.model_dump() for session in sessions],
            "total": total,
            "limit": limit,
            "offset": offset
        }
    except Exception as e:
        logger.error(f"Failed to get training sessions: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.get("/training-session/{session_id}")
async def get_training_session_status(
    session_id: int = Path(..., description="Training session ID")
):
    """Get training session status"""
    try:
        session = await text2sql_service.get_training_session_status(session_id)
        if not session:
            raise HTTPException(status_code=404, detail="Training session not found")
        
        return session.model_dump()
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to get training session status: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/embeddings")
async def generate_embeddings(
    datasource_id: int = Body(..., description="Datasource ID"),
    training_data_ids: List[int] = Body(..., description="Training data IDs to process")
):
    """Generate embeddings for specific training data"""
    try:
        result = await text2sql_service.generate_embeddings_for_training_data(
            datasource_id=datasource_id,
            training_data_ids=training_data_ids
        )
        return result
    except Exception as e:
        logger.error(f"Failed to start embedding generation: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


# Statistics and Utility Endpoints

@router.get("/statistics", response_model=Text2SQLStatistics)
async def get_statistics(
    datasource_id: Optional[int] = Query(None, description="Filter by datasource ID")
):
    """Get Text2SQL statistics"""
    try:
        stats = await text2sql_service.get_statistics(datasource_id)
        return stats
    except Exception as e:
        logger.error(f"Failed to get statistics: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.get("/datasource/{datasource_id}/training-summary")
async def get_datasource_training_summary(
    datasource_id: int = Path(..., description="Datasource ID")
):
    """Get training data summary for a specific datasource"""
    try:
        summary = await text2sql_service.get_datasource_training_summary(datasource_id)
        return summary
    except Exception as e:
        logger.error(f"Failed to get training summary: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/cleanup")
async def cleanup_old_data(
    days_to_keep: int = Body(30, ge=1, le=365, description="Number of days to keep data")
):
    """Start cleanup task for old data"""
    try:
        result = await text2sql_service.cleanup_old_data(days_to_keep)
        return result
    except Exception as e:
        logger.error(f"Failed to start cleanup task: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


# WebSocket Endpoint

@router.websocket("/ws/{datasource_id}")
async def websocket_endpoint(websocket: WebSocket, datasource_id: int):
    """WebSocket endpoint for real-time Text2SQL updates"""
    await handle_websocket_connection(websocket, datasource_id)


# Training Endpoints - DDL, Documentation, SQL

@router.post("/train-ddl")
async def train_ddl(
    datasource_id: int = Body(..., description="Datasource ID"),
    auto_extract: bool = Body(False, description="Auto extract database schema"),
    database_name: Optional[str] = Body(None, description="Database name"),
    ddl_content: Optional[str] = Body(None, description="DDL content"),
    skip_existing: bool = Body(True, description="Skip existing tables")
):
    """Train model with DDL (Data Definition Language) statements"""
    try:
        result = await text2sql_service.train_ddl(
            datasource_id=datasource_id,
            auto_extract=auto_extract,
            database_name=database_name,
            ddl_content=ddl_content,
            skip_existing=skip_existing
        )
        return result
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to train DDL: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/train-documentation")
async def train_documentation(
    datasource_id: int = Body(..., description="Datasource ID"),
    documentation: str = Body(..., description="Documentation content"),
    doc_type: str = Body("general", description="Documentation type"),
    metadata: Optional[Dict[str, Any]] = Body(None, description="Additional metadata")
):
    """Train model with documentation content"""
    try:
        result = await text2sql_service.train_documentation(
            datasource_id=datasource_id,
            documentation=documentation,
            doc_type=doc_type,
            metadata=metadata or {}
        )
        return result
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to train documentation: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/train-sql")
async def train_sql(
    datasource_id: int = Body(..., description="Datasource ID"),
    sql_pairs: List[Dict[str, str]] = Body(..., description="SQL question-answer pairs")
):
    """Train model with SQL question-answer pairs"""
    try:
        result = await text2sql_service.train_sql_pairs(
            datasource_id=datasource_id,
            sql_pairs=sql_pairs
        )
        return result
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to train SQL pairs: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


# Advanced Features - Question Answer Mode

@router.post("/answer", response_model=QuestionAnswerResponse)
async def answer_question(request: QuestionAnswerRequest):
    """Answer natural language question with SQL generation and optional execution"""
    try:
        response = await text2sql_service.answer_question(request)
        return response
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to answer question: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/batch-train")
async def batch_train_sql_pairs(request: BatchTrainingRequest):
    """Batch training with SQL question-answer pairs"""
    try:
        result = await text2sql_service.batch_train_sql_pairs(request)
        return result
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to batch train SQL pairs: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/generate-embeddings")
async def generate_embeddings(
    datasource_id: int = Body(..., description="Datasource ID"),
    training_data_ids: List[int] = Body(None, description="Specific training data IDs (optional)")
):
    """Generate embeddings for training data"""
    try:
        if training_data_ids:
            # Generate embeddings for specific training data
            result = await text2sql_service.generate_embeddings_for_training_data(
                datasource_id=datasource_id,
                training_data_ids=training_data_ids
            )
        else:
            # Generate embeddings for all training data of this datasource
            training_data, _ = await text2sql_service.get_training_data_list(
                datasource_id=datasource_id,
                limit=1000  # Get all training data
            )

            if not training_data:
                return {
                    "success": False,
                    "message": "No training data found for this datasource",
                    "task_id": None
                }

            result = await text2sql_service.generate_embeddings_for_training_data(
                datasource_id=datasource_id,
                training_data_ids=[data.id for data in training_data]
            )

        return result
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to generate embeddings: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


@router.post("/retrain-model")
async def retrain_model(
    datasource_id: int = Body(..., description="Datasource ID"),
    force_rebuild: bool = Body(False, description="Force rebuild all embeddings")
):
    """Retrain model for a datasource"""
    try:
        task_id = await text2sql_service.retrain_model(
            datasource_id=datasource_id,
            force_rebuild=force_rebuild
        )
        return {
            "success": True,
            "message": "Model retraining started",
            "task_id": task_id
        }
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"Failed to start model retraining: {e}")
        raise HTTPException(status_code=500, detail="Internal server error")


# Health Check Endpoint

@router.get("/health")
async def health_check():
    """Health check endpoint"""
    try:
        health_status = await text2sql_service.health_check()
        return health_status
    except Exception as e:
        logger.error(f"Health check failed: {e}")
        return {
            "status": "unhealthy",
            "service": "text2sql",
            "error": str(e)
        }

File: web/src/core/api/text2sql.ts
/**
 * Text2SQL API client for Olight
 *
 * Provides TypeScript interfaces and API functions for Text2SQL functionality
 * including SQL generation, training data management, and real-time monitoring.
 */

import { resolveServiceURL } from "./resolve-service-url";

// Types and Interfaces
export interface SQLGenerationRequest {
  datasource_id: number;
  question: string;
  include_explanation?: boolean;
  embedding_model_id?: number;
}

export interface SQLGenerationResponse {
  query_id: number;
  question: string;
  generated_sql: string;
  explanation?: string;
  confidence_score: number;
  similar_examples: Array<{ query: string; usage_count: number }>;
  generation_time_ms: number;
}

export interface SQLExecutionRequest {
  query_id: number;
  limit?: number;
}

export interface SQLExecutionResponse {
  query_id: number;
  status: 'PENDING' | 'SUCCESS' | 'FAILED';
  result_data?: Array<Record<string, any>>;
  result_rows?: number;
  execution_time_ms?: number;
  error_message?: string;
}

export interface TrainingDataRequest {
  datasource_id: number;
  content_type: 'SQL' | 'SCHEMA' | 'DOCUMENTATION' | 'DDL';
  content: string;
  question?: string;
  sql_query?: string;
  table_names?: string[];
  database_schema?: Record<string, any>;
  metadata?: Record<string, any>;
}

export interface TrainingDataResponse {
  id: number;
  datasource_id: number;
  content_type: string;
  question?: string;
  sql_query?: string;
  content: string;
  table_names?: string[];
  database_schema?: Record<string, any>;
  metadata?: Record<string, any>;
  content_hash: string;
  is_active: boolean;
  is_validated: boolean;
  validation_score?: number;
  created_at: string;
  updated_at: string;
}

export interface QueryHistory {
  id: number;
  user_question: string;
  generated_sql: string;
  datasource_id: number;
  status: string;
  execution_time_ms?: number;
  result_rows?: number;
  confidence_score?: number;
  model_used?: string;
  explanation?: string;
  created_at: string;
}

export interface TrainingSession {
  id: number;
  datasource_id: number;
  session_name?: string;
  model_version?: string;
  status: 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED';
  training_data_count?: number;
  accuracy_score?: number;
  validation_score?: number;
  training_time_seconds?: number;
  started_at: string;
  completed_at?: string;
  notes?: string;
}

export interface Text2SQLStatistics {
  total_queries: number;
  successful_queries: number;
  failed_queries: number;
  average_confidence?: number;
  total_training_data: number;
  training_data_by_type: Record<string, number>;
  last_query_time?: string;
  last_training_time?: string;
}

// Advanced Features Types
export interface QuestionAnswerRequest {
  question: string;
  datasource_id: number;
  execute_sql?: boolean;
  format_result?: boolean;
  include_explanation?: boolean;
  embedding_model_id?: number;
}

export interface QuestionAnswerResponse {
  question: string;
  generated_sql: string;
  explanation?: string;
  confidence_score: number;
  execution_result?: SQLExecutionResponse;
  formatted_answer?: string;
  generation_time_ms: number;
}

export interface SQLPair {
  question: string;
  sql: string;
  explanation?: string;
}

export interface BatchTrainingRequest {
  datasource_id: number;
  database_name?: string;
  sql_pairs: SQLPair[];
  embedding_model_id?: number;
  overwrite_existing?: boolean;
}

export interface BatchTrainingResponse {
  added_count: number;
  skipped_count: number;
  total_pairs: number;
  errors: string[];
  success_rate: number;
}

export interface WebSocketMessage {
  type: 'training_progress' | 'training_completed' | 'training_failed' | 'query_execution' | 'system_status' | 'error';
  datasource_id?: number;
  session_id?: number;
  task_id?: string;
  progress?: number;
  message?: string;
  details?: Record<string, any>;
  timestamp: string;
}

// API Functions
const API_BASE = 'text2sql';

export const text2sqlApi = {
  // SQL Generation and Execution
  async generateSQL(request: SQLGenerationRequest): Promise<SQLGenerationResponse> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/generate`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to generate SQL');
    }

    return response.json();
  },

  async executeSQL(request: SQLExecutionRequest): Promise<SQLExecutionResponse> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/execute`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to execute SQL');
    }

    return response.json();
  },

  // Query History
  async getQueryHistory(params: {
    datasource_id?: number;
    status?: string;
    limit?: number;
    offset?: number;
  } = {}): Promise<{ queries: QueryHistory[]; total: number; limit: number; offset: number }> {
    const searchParams = new URLSearchParams();
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined) {
        searchParams.append(key, value.toString());
      }
    });
    
    const response = await fetch(resolveServiceURL(`${API_BASE}/history?${searchParams}`));

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to get query history');
    }

    return response.json();
  },

  // Training Data Management
  async addTrainingData(request: TrainingDataRequest): Promise<TrainingDataResponse> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/training`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to add training data');
    }

    return response.json();
  },

  async getTrainingData(params: {
    datasource_id?: number;
    content_type?: string;
    is_active?: boolean;
    limit?: number;
    offset?: number;
  } = {}): Promise<{ training_data: TrainingDataResponse[]; total: number; limit: number; offset: number }> {
    const searchParams = new URLSearchParams();
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined) {
        searchParams.append(key, value.toString());
      }
    });
    
    const response = await fetch(resolveServiceURL(`${API_BASE}/training?${searchParams}`));

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to get training data');
    }

    return response.json();
  },

  async updateTrainingData(trainingId: number, request: TrainingDataRequest): Promise<TrainingDataResponse> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/training/${trainingId}`), {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to update training data');
    }

    return response.json();
  },

  async deleteTrainingData(trainingId: number, softDelete: boolean = true): Promise<void> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/training/${trainingId}?soft_delete=${softDelete}`), {
      method: 'DELETE',
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to delete training data');
    }
  },

  // Training Sessions and Model Management
  async retrainModel(datasourceId: number, forceRebuild: boolean = false): Promise<{ task_id: string; message: string }> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/retrain/${datasourceId}?force_rebuild=${forceRebuild}`), {
      method: 'POST',
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to start retraining');
    }

    return response.json();
  },

  async startTrainingSession(request: {
    datasource_id: number;
    session_name?: string;
    model_version?: string;
    training_parameters?: Record<string, any>;
    notes?: string;
  }): Promise<{ session_id: number; task_id: string; status: string }> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/training-session`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to start training session');
    }

    return response.json();
  },

  async getTrainingSessions(params: {
    datasource_id?: number;
    limit?: number;
    offset?: number;
  } = {}): Promise<{ sessions: TrainingSession[]; total: number; limit: number; offset: number }> {
    const searchParams = new URLSearchParams();
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined) {
        searchParams.append(key, value.toString());
      }
    });
    
    const response = await fetch(resolveServiceURL(`${API_BASE}/training-sessions?${searchParams}`));

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to get training sessions');
    }

    return response.json();
  },

  async getTrainingSessionStatus(sessionId: number): Promise<TrainingSession> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/training-session/${sessionId}`));

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to get training session status');
    }

    return response.json();
  },

  // Statistics and Utilities
  async getStatistics(datasourceId?: number): Promise<Text2SQLStatistics> {
    const params = datasourceId ? `?datasource_id=${datasourceId}` : '';
    const response = await fetch(resolveServiceURL(`${API_BASE}/statistics${params}`));

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to get statistics');
    }

    return response.json();
  },

  async getTrainingSummary(datasourceId: number): Promise<{
    total_items: number;
    total_validated: number;
    by_type: Record<string, { count: number; validated_count: number; avg_validation_score?: number }>;
  }> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/datasource/${datasourceId}/training-summary`));

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to get training summary');
    }

    return response.json();
  },

  async generateEmbeddings(request: {
    datasource_id: number;
    training_data_ids?: number[];
  }): Promise<{ task_id: string; status: string; success: boolean; message?: string }> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/generate-embeddings`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to start embedding generation');
    }

    return response.json();
  },

  async cleanupOldData(daysToKeep: number = 30): Promise<{ task_id: string; status: string }> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/cleanup`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ days_to_keep: daysToKeep }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to start cleanup');
    }

    return response.json();
  },

  // WebSocket Connection
  createWebSocket(datasourceId: number): WebSocket {
    const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
    // For WebSocket, we need to construct the URL differently
    const baseUrl = resolveServiceURL(`${API_BASE}/ws/${datasourceId}`);
    const wsUrl = baseUrl.replace(/^http/, 'ws');
    return new WebSocket(wsUrl);
  },

  // Advanced Features - Question Answer Mode
  async answerQuestion(request: QuestionAnswerRequest): Promise<QuestionAnswerResponse> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/answer`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to answer question');
    }

    return response.json();
  },

  async batchTrainSQLPairs(request: BatchTrainingRequest): Promise<BatchTrainingResponse> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/batch-train`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to batch train SQL pairs');
    }

    return response.json();
  },

  // DDL Training
  async trainDDL(datasourceId: number, request: {
    auto_extract?: boolean;
    database_name?: string;
    ddl_content?: string;
    embedding_model_id?: number;
    skip_existing?: boolean;
  }): Promise<{
    success: boolean;
    message?: string;
    error?: string;
    total_items?: number;
    successful_items?: number;
    failed_items?: number;
    training_time?: number;
    details?: any[];
  }> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/train-ddl`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ datasource_id: datasourceId, ...request }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to train DDL');
    }

    return response.json();
  },

  // Documentation Training
  async trainDocumentation(datasourceId: number, request: {
    documentation: string;
    doc_type: 'business' | 'technical' | 'api' | 'other';
    embedding_model_id?: number;
    metadata?: Record<string, any>;
  }): Promise<{
    success: boolean;
    message?: string;
    error?: string;
    total?: number;
    successful?: number;
    failed?: number;
  }> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/train-documentation`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ datasource_id: datasourceId, ...request }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to train documentation');
    }

    return response.json();
  },

  // SQL Training
  async trainSQL(datasourceId: number, request: {
    sql_pairs: Array<{ question: string; sql: string }>;
    embedding_model_id?: number;
  }): Promise<{
    success: boolean;
    message?: string;
    error?: string;
    total?: number;
    successful?: number;
    failed?: number;
  }> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/train-sql`), {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ datasource_id: datasourceId, ...request }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to train SQL');
    }

    return response.json();
  },

  // Health Check
  async healthCheck(): Promise<{ status: string; service: string; timestamp: string }> {
    const response = await fetch(resolveServiceURL(`${API_BASE}/health`));

    if (!response.ok) {
      throw new Error('Text2SQL service is not healthy');
    }

    return response.json();
  },
};
